File: 374-guess-number-higher-or-lower/374.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        if n == 1:
            return 1

        low = 1
        high = n
        pick = 1

        while pick <= n:
            next_pick = (low + high) // 2
            if next_pick == pick:
                pick += 1
            else:
                pick = next_pick
            x = guess(pick)
            if x == 0:
                return pick
            if x == -1:
                high = pick
            elif x == 1:
                low = pick

        return pick

File: 374-guess-number-higher-or-lower/374.c
/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */

int guessNumber(int n){
    unsigned int left = 1;
    unsigned int right = n;
    while (left < right) {
        unsigned long long mid = (left + right) / 2;
        int ret = guess(mid);
        if (ret == 0)
            return mid;
        else if (ret == -1)
            right = mid;
        else if (ret == 1)
            left = mid + 1;
    }

    return right;
}
